tracer/src/Datadog.Trace/Debugger/Instrumentation/Collections/ProbeDataCollection.cs

```diff
@@ -42,52 +42,84 @@ internal static ProbeMetadataCollection Instance
         }
 
         /// <summary>
-        /// Tries to create a new <see cref="ProbeData"/> at <paramref name="index"/>.
+        /// Tries to create a new <see cref="ProbeData"/> at <paramref name="index"/> and return it.
         /// </summary>
         /// <param name="index">The index of the probe inside <see cref="EverGrowingCollection{TPayload}.Items"/></param>
         /// <param name="probeId">The id of the probe/></param>
         /// <returns>true if succeeded (either existed before or just created), false if fails to create</returns>
-        public bool TryCreateProbeMetadataIfNotExists(int index, string probeId)
+        public ref ProbeData TryCreateProbeDataIfNotExists(int index, string probeId)
         {
-            if (IsIndexExists(index))
+            if (IsIndexExists(index, probeId))
             {
-                return true;
+                // Between the time `IsIndexExists` was called and the time `GetProbeDataIndex` will be called there
+                // might be context switches that will invalidate the ProbeData located at `index`.
+                // So we essentially perform the ProbeId checking twice to make sure the `ProbeData` returned is indeed related
+                // to the probeId given. Of course, one tick later it could be invalidated too, but at least we willreturn `ProbeData` that
+                // is related to the given `probeId` and not of another probe.
+                return ref GetProbeDataAtIndex(index, probeId);
             }
 
             // Create a new one at the given index
             lock (ItemsLocker)
             {
-                if (IsIndexExists(index))
+                if (IsIndexExists(index, probeId))
```

Comment:
`IndexExists`

Reply:
Done in 90e0ba24cfd43eab1fca79e8a887db33e9ddf49e.